Improvements to wallet syncing and monitoring

AdamISZ opened the pull request:
Idea of this PR was prompted by some related problems.

See #328 - there seems to be some flaw in full sync at least (and fast also perhaps) involving not recognising/registering new addresses to the wallet such that restarts almost always involve retrying the sync when that shouldn't be necessary.
Not obviously related, but #336 for example is a simple failure at UI level where wallet does not dynamically update the GUI; similar but more fundamental, a deposit into the wallet during operation is not recognized until restart.
I think all these should be fixed, and it makes sense to do it as one "project". I'd like to achieve these things:

* Make wallet sync *aggressive* and *optimistic* and *correct* - so iron out any bugs, go with the approach used in `--fast` sync of assuming "normal" conditions and make it *always* work, first time, then - for the restarting-on-new-Core-wallet case, we can still keep the more clunky fallback.
* Create a better/sharper interface between the client apps and the blockchain *via* the wallet. So that only the wallet will be affected by future changes (I'm thinking particularly of the possible future BIP157/8.
usage)
* Better asynchronous update. Let the clients (including the GUI as another type of client) subscribe to events, in particular changes to the wallet's utxo set. This can be made generic so the way it's done to solve the above mentioned "dynamic deposit" case works the same as for taker and makers watching for completions of coinjoins.


The first pushed commit here is something fairly minor, but part of the clean up: removed duplicated `import_new_addresses` functions and replaced them with an atomically in-wallet call, so that we can "guarantee" that all addresses used during JM operation have been imported.

AdamISZ commented:
As of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/82406550e986b361941462d18eb4d3ab5baa1bfd I can report that I no longer have the problem of "need to do multiple restarts" from my ongoing mainnet testing.

The change is explained in some detail in the commit message.
However I did neglect to mention one other minor point there, so I'll summarise here:

(1) save the wallet indices at the end of a successful sync, as otherwise a change in the index will not be persisted unless a transaction happens (meaning on restart, you have to go back to the previous, unsynced index).
(2) This is less obvious, but I decided to include a step of importing from (current index) to (current index + gaplimit) *each* time we request a new address. This is very cheap and it doesn't harm anything to import an address multiple times, and it avoids the problem that was causing unnecessarily-forced-restarts (see #328 - this specific commit should fix that), which was caused by the previous "import the new address" being insufficient for the sync, which requires everything imported up to the gap limit. Hope that's clear.

@kristapsk if you have a chance to test, let me know if this commit fixes sync for you on whatever big wallet you have around.

@undeath This isn't much more than a curiosity at this point, but don't you think [these](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L596-L597) lines are un-needed?

kristapsk commented:
https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/files#diff-79627d21bc41aff521fe45af1ee3968aR560 breaks wallet-tools' history method and probably other read-only stuff, raises `jmclient.storage.StorageError: Read-only storage cannot be saved.`

AdamISZ commented:
@kristapsk yep i saw it earlier today also. Fix is in the next commit.

Basically, some wallet-tool methods are "noscan" and that's important so you can access the data without trying to sync the wallet (like `showseed`). We also had some additional methods marked 'read-only'; in my judgement it's better to not have that, so that the wallet's index (or sometimes called index_cache) can be updated after a sync, because otherwise if you run wallet-tool multiple times it'll have to redo all the same scanning work every time to update the index to the correct value.

So I removed the 'read-only' flag from runs of wallet-tool on the methods 'display' and similar (like 'showutxos') that must sync/scan the wallet/blockchain.

I judge there's no meaningful loss in security; this is not a long-running process like the maker or taker. You're still reading the secrets into memory, so there's no point blocking an update to the wallet file itself.

kristapsk commented:
There is big downside with removing read-only flag - now you can't run them while yield generator is running, because of wallet locks. So, it requires to shutdown yg, just to view history or list addresses (let's say, you want to send additional funds to your yg wallet) or do anything else with a specific wallet.

AdamISZ commented:
Ah gotcha. Will figure it out tomorrow then.

AdamISZ commented:
OK I'll just remove that index-saving code, it won't help anyway in normal conditions, it would only rarely make things more efficient, and I agree it's better not to screw up the locking function vs allowing read-only query.

kristapsk commented:
With 1e4e63c67f3980cf0707380f8bb6ba1887418c91 haven't found any new issues and it solves #328 for me.

undeath commented:
> @undeath This isn't much more than a curiosity at this point, but don't you think [these](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L596-L597) lines are un-needed?

Those are an emergency fallback in case the wallet has been uncleanly closed (eg because of a crash). Removing those could cause jm to not recognise already used addresses, re-using them instead (and failing to display funds in there). Removing those would mean the user has no way to instruct jm to scan addresses beyond the wallet-saved index.

kristapsk commented:
I am running this all the time and didn't got any "restart Bitcoin Core..." messages until today. Only thing different is that I deposited some funds into external address of a wallet. But I think I have done it before (like in a last month or so) too, but don't remember having that message. In any case, clearly an improvement, although if depositing funds can cause it, will not be ok for #336.

AdamISZ commented:
> I am running this all the time and didn't got any "restart Bitcoin Core..." messages until today. Only thing different is that I deposited some funds into external address of a wallet. But I think I have done it before (like in a last month or so) too, but don't remember having that message. In any case, clearly an improvement, although if depositing funds can cause it, will not be ok for #336.

Well that's a disappointment. Clearly the intention was to prevent this; I'll try to reproduce, but pretty sure this was exactly the scenario I tested before (and indeed, both in testing and real world use I am not seeing this), so I don't know what causes it. If you can provide any more info on the circumstances it may help.

I'll get back on working on this whole attempted syncing revamp today, or this week, at least I'll try to make progress ...

kristapsk commented:
> deposit while JM is running, which is distinct I guess, from a deposit between runs

I was depositing while yg is running and then got error with wallet-tool, IIRC.

Will be mosly offline for about a week now, will be able to do more testing from my side next week.

AdamISZ commented:
So I've spent a day or more reviewing and analysing the syncing code as-is, and I've come to the following conclusions:

* The example you give, @kristapsk , is valid (and I can also confirm in regtest); an external deposit will always cause this problem (not sure how I thought this wasn't the case before but it is).
* The specific reasons for it are related to the assumption-free mode of operation of the default (non-fast) syncing.
* This problem does not occur with the `fast` sync option. More on this below.

So, the result of the review and discussion is that I believe I made an error in not promoting the `fast` sync option, back at the time (years ago), to the default method of syncing. As the name suggests, it is indeed fast, as in a second or two even for very large/old wallets, but it is also more reliable - if certain, reasonable assumptions hold.

The approach behind fast sync:

We work on the assumption that all addresses that were ever used in the wallet label `joinmarket-XX` were those created by that Joinmarket wallet. Based on this, we go through these steps:
* Collect all addresses ever used on that label with rpc `listaddressgroupings`
* Additionally, record the current wallet index values per branch as per the wallet file, if they exist
* Working in batches, search forward through the HD wallet branches, collecting instances of usage of addresses found in step 1.
* Stop searching forward when all addresses identified in step 1 were matched to addresses in the JM HD wallet
* Once finished, mark the wallet indices as the highest found used address or the current wallet indexes (from step 2), whichever is higher.

Notice how this leverages two basic assumptions - no "foreign" usage of the JM HD wallet's specific label, and the wallet branches are not used beyond batch size * num iterations - to make a very simple algorithm. With perhaps a bump to the latter (batch size maybe 200 should be fine), this should cover every situation where the wallet is created on one Core instance and then progressively used.

About imports:
A primary source of complexity is imports required on the 'lower level' Core wallet. In case of progressive usage on one Core instance, we only need to ensure that imports are effected up to the gap limit beyond whatever is the current index, which is now done with https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/1e4e63c67f3980cf0707380f8bb6ba1887418c91 . So any deposit by a user to an address which is actually displayed would be an address that's already imported. This addresses @kristapsk 's issue (note: it *doesn't* address it with the current default sync mode, see below).

About the detailed/non-fast sync:

The basic principle here is not to assume anything; it is specifically designed to address the needs of a recovery operation of a wallet on a new Core instance, where there is no wallet jmdat file and thus no JM HD wallet index, and there are no imports. It also does not assume that there are no usages of addresses beyond the gap limit, nor does it assume (or, let's say, *use* the assumption) that there are no usages of addresses at that label that are "foreign" to the JM HD wallet.

This set of assumptions, or rather non-assumptions both makes it slow and also makes it hard to reason about and unreliable. One example of that is what @kristapsk found above: when he deposited to a new address, it created a situation where the set of "gap addresses" (the gap_limit addresses on each branch beyond the currently set index) was bumped forwards and then included new, un-imported addresses, causing the triggering of [this](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L542) line, which shuts down after doing address imports, specifically because it does not 100% know that the newly imported addresses are unused (in fact, that assumption is *nearly* safe, so one could argue for changing this), and if they are used, then a rescan is needed.

The more general problematic symptom, which existed before https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/1e4e63c67f3980cf0707380f8bb6ba1887418c91 , was that restarts were needed after every transaction in the wallet. This was a very similar but not identical situation to the above: new transactions caused new imports and bumped the index, but the imports triggered did not include the whole of the gap addresses beyond that current index, and so this triggered a quit even earlier, [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L526).

In case it wasn't clear, the only difference between these two situations was that the "deposit" scenario didn't bump the index, whereas a transaction did.

This is just a subset of the complexity here. It's important to note that not only is the second algorithm more complex, but that *because it makes fewer assumptions it actually can often fail where the first algorithm succeeds, as well as being slower*.

Proposed changes:

Make `--fast` the default. Note that it handles initial startup perfectly fine, see [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L446-L454). If a person continues to use Joinmarket normally on one Core instance, they should never need to do anything else, until they hit the limits hardcoded as batch_size and number of iterations (with the bump to 200 mentioned above, this should be at about 4000 addresses per branch). In any case, we can of course make sure that the error message clearly indicates:

* If the batch limits are exceeded, clearly state it and suggest making a new wallet or giving an option to bump
* Important message displayed prominently, always: if the balance shown is 0 or less than expected, you should try to sync in "recovery" mode

Then also: rename the current default sync mode to "recovery" mode, as that will be its primary use case.

All of the above is about sync, not about wallet *monitoring* which should be addressed separately.

Ping @undeath @adlai @chris-belcher @AlexCato just in case you have thoughts on this.

AlexCato commented:
> Proposed changes:
> 
> Make `--fast` the default. Note that it handles initial startup perfectly fine, see [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/blockchaininterface.py#L446-L454). If a person continues to use Joinmarket normally on one Core instance, they should never need to do anything else, until they hit the limits hardcoded as batch_size and number of iterations (with the bump to 200 mentioned above, this should be at about 4000 addresses per branch). In any case, we can of course make sure that the error message clearly indicates:
> 
>     * If the batch limits are exceeded, clearly state it and suggest making a new wallet or giving an option to bump
> 
>     * Important message displayed prominently, always: if the balance shown is 0 or less than expected, you should try to sync in "recovery" mode
> 
> 
> Then also: rename the current default sync mode to "recovery" mode, as that will be its primary use case.


I approve of all these changes and would be happy to see them. I guess most users are unaware of --fast and barely read any documentation. I've lately forgot --fast sometimes, and boy that does take forever on a wallet that has seen some use. So it's a very userfriendly change to make --fast the default; also renaming the other one to recovery is good.

AdamISZ commented:
This was force-pushed today. There is already here a major refactoring, and I am increasingly confident that it's a very useful change. Since it's very big, here's a summary:

* About 400 LOC removed from blockchaininterface. They are: (1) the various collect_addresses* and rewind_indices* functions which are moved into the wallet, as they are pure wallet functions, and (2) removal of 3 very large functions which were used to monitor transactions (`add_tx_notify`, `tx_watcher` and `outputs_watcher`). For replacement of those, see:
* New module in `jmclient` called `wallet_service.py` which houses an object `WalletService`. This has a couple of functions: to encapsulate access to the wallet (theoretically it could allow use of multiple wallets simultaneously, even) and better define an interface, and also encapsulates *some* access to the blockchaininterface, although currently not all. More importantly though, perhaps, is having a central monitoring loop for transactions that can be effectively subscribed to (see `register_callbacks` comments) to get updated on new transactions and confirmations, as before, but in a slightly less complex way. This architecture has lessened the rather painfully twisty cross referencing from client to bci to wallet, back to bci etc etc.
* Further: **Now clients (Takers, Makers) do not need to manually call add/remove utxos, this happens automatically within the wallet service, i.e. the utxo set is updated as part of the transaction monitoring loop which is always-on.** This last point is crucial because it makes it dead simple to keep the Qt in-sync with any external transactions (deposits etc.) as is kind of required for a decent GUI wallet. (also long running ygs should be able to use deposits, same principle).
* Some changes to the wallet in line with the above - a new method `process_new_tx` encapsulates the above, so that the utxo set is always updated with transactions that are relevant to the wallet (and any that are not are just ignored).
* Access to the wallet methods is now almost exclusively via the WalletService interface, even where this means trivial pass-through functions (I'm currently vacillating on whether that should be 100% strict - probably, yes - but we can always change it later since it's just refactoring).

Avoiding getting lost in detail by listing more, but next steps:
* Main CLI scripts working, but need to update the lesser CLI tools.
* Update joinmarket-qt.py, including dynaming refresh of wallet balance
* Update tests

None of these should be particularly hard at all, but they'll take some days. I'll also need to rebase of course. I expect this to be done within a week, probably. Feedback would be appreciated, especially if you can test *after* I've fixed the tests (dubious value to test it now).

AdamISZ commented:
Syncing:

The above commit adds the "default to fast sync" as previously discussed. Otherwise, the syncing methods are largely unchanged (except that you don't have to call them directly in user-level scripts; the `WalletService` does that for you).

I should also add that these changes have not addressed recently discovered problems with regard to imports, but they provide a route to doing so - we can add periodically woken up calls to do imports in small batches (not exactly a beautiful thing, but we seem to have no choice given what is effectively a bug in Core for slow HDs).

AdamISZ commented:
**I have removed this from WIP status** because I think a natural end point has been reached that achieves these basic goals (not entirely what I set out, but a very significant chunk of improvement):

* Both yield generators and GUI will automatically recognize deposits (and in the latter case, show them); note how this happens now without code explicitly, because the wallet is constantly updated via the WalletService.
* Fast sync is now default, which will take away all the problems for normal operations afaict
* The wallet service architecture just uses one transaction monitoring loop, and much of the complexity in the blockchain interface code has been removed.
* The wallet is now almost entirely encapsulated within the wallet service, which could lead to smarter ways of handling wallets in future (including making entirely new wallet types with no impact at higher levels.

There are other minor things, notably the Qt Wallet tab properly respecting expansion states, and a few small refactorings.

What has **not** been done is to cache importing, but I believe this will be much easier with the WalletService architecture.

* Testing

What has been tested:
Apart from the basic test suite, I have tested all of these on regtest:
wallet-tool display, showutxos, sync, showseed, generate
tumbler basic runs
sendpayment direct send and join
Qt: wallet creation and syncing, deposits
yieldgenerator funcion

TODO: keep testing as many functions as possible. I will test payjoin (although it runs in functional tests), and I will keep running from Maker and Taker on mainnet. So far I have run yieldgenerators and seen transactions go through, but it needs quite a bit more.

So I would very much appreciate people running this in various ways, especially on mainnet. Since it's a broad change it needs a fair bit, there will doubtless be one or two more bugs, but I expect them to be minor at this point.

(Will squash at end but left commits for now in case it's easier to read)

undeath commented:
This looks like a significant improvement over the kind of syncing done right now, thanks for this! I plan on reviewing and testing this code in the near future.

When I start jm I see a bunch of "saw tx on network" messages with this code, which is a little confusing.

AdamISZ commented:
Thanks.

> When I start jm I see a bunch of "saw tx on network" messages with this code, which is a little confusing.

Yes you're right. I haven't got round to it, it's just an artifact of setting up the monitoring loop by "caching" the current set of returns from `listtransactions`. Will do.

undeath commented on jmclient/jmclient/yieldgenerator.py:
```diff
@@ -144,7 +143,7 @@ def on_tx_unconfirmed(self, offer, txid, removed_utxos):
         return [], [newoffers[0]]
 
-    def on_tx_confirmed(self, offer, confirmations, txid):
+    def on_tx_confirmed(self, offer, txid):
```
This should have the same required arguments as the parent class `Maker`.

AdamISZ replied:
OK so TIL that `@abc.abstractmethod` doesn't check anything other than a name! So that's a small gotcha for me at least.

Indeed both the function signatures for the `on_tx_[un]confirmed` variants had gone out of whack. It's additionally confusing that there are asymmetries between Taker and Maker here that I would forgive anyone for finding obtuse, but in any case I have fixed them up to actually match the base class signature as of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/820fc11cad9008a2ad50065bb3d0bb495b9809d2 (note the P2EP maker uses the function slightly differently).

undeath commented on jmclient/jmclient/client_protocol.py:
```diff
@@ -283,52 +293,48 @@ def unconfirm_callback(self, txd, txid):
-        jlog.info('saw tx on network, removed_utxos=\n{}'.format('\n'.join(
-            '{} - {}'.format(u, fmt_tx_data(tx_data, self.client.wallet))
-            for u, tx_data in removed_utxos.items())))
+        if offerinfo is None:
```
Logging this case would be beneficial to aid debugging. The line could be simplified to `else:` instead of the `if` check.

AdamISZ replied:
Not really, I think.
The rather big redesign here is that we can call the unconfirm callback lots of times on transactions which are not the one we're tracking (mostly that won't actually happen, but it can, and not only for joins in parallel); if it's not the one we're tracking, we return False so we know to keep looking for it, and only return True when we get a match. I don't think it's useful to log non-matches as there can be tons of them and they don't tell us much (except in a development phase).

The point about `else` is right though (I should get used to that!), 'fixed' in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/e645104391bd07290d31e4ee9f6b6cd3f32c74a4

undeath replied:
I see, I was of the impression this is mostly an error condition. Maybe still and a `jlog.debug` just in case?

AdamISZ replied:
I mean you could do it for testing, but I don't want to log every time a callback gets fired on an inapplicable transaction.

undeath commented on jmclient/jmclient/client_protocol.py:
```diff
@@ -283,52 +293,48 @@ def unconfirm_callback(self, txd, txid):
+        return True
 
-    def confirm_callback(self, txd, txid, confirmations):
+    def confirm_callback(self, txd, txid, confirms):
```
this whole code is mostly duplicating `unconfirm_callback`

AdamISZ replied:
Refactored the transaction matching to be common ( https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/eac7fd544b6728cf1753d935927f2148f738c32d ) , but left the other (callRemote) in place, seems pretty minor.

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+    preserve encapsulation.
+    """
+
+    def get_txtype(self):
```
All those wallet pass-through methods in `WalletService` could be replaced with a single `__getattr__ ` implementation, that simply calls the wallet whenever an unknown method is called on the `WalletService`. This will make the interface much easier to maintain.

AdamISZ replied:
Oh cool idea, thanks.

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/248dbbad63bd27862e7b3b9354468db13d09aec3

undeath commented on jmclient/jmclient/taker.py:
```diff
@@ -820,22 +828,44 @@ def self_sign_and_push(self):
+        return True
+
     def unconfirm_callback(self, txd, txid):
+        if not self.tx_match(txd):
```
under what circumstances can this happen?

AdamISZ replied:
Same as above discussion, each tx fires the extant callbacks, so we check if it matches. In Taker this must be exactly `self.latest_tx` not a dict since Taker processes only in series.

undeath commented on jmclient/jmclient/wallet.py:
```diff
@@ -566,6 +537,10 @@ def remove_old_utxos(self, tx):
+        if len(removed_utxos.keys()) > 0:
+            jlog.info('saw tx on network, removed_utxos=\n{}'.format('\n'.join(
+            '{} - {}'.format(u, fmt_tx_data(tx_data, self))
+            for u, tx_data in removed_utxos.items())))
```
It seems a bit weird to put this in the wallet code. I'd more expect the caller to emit such a message.

AdamISZ replied:
Indeed it does seem a bit weird, but the point is there isn't a caller, in a sense: since we're moving to a reactive mode for transactions hitting the wallet, this was not triggered by any action in Joinmarket code, only by what the blockchain told us. What is perhaps illogical is: we have this message for removals but not additions. On balance I think I'd prefer both. We could also put these messages in wallet service but it just seemed logical to keep this message (which already existed) but put it where it happens. What do you think?

undeath replied:
The thing is, this says something about network activity, but ultimately the wallet knows nothing about where its information is sourced from. So this log message should be somewhere closer to the actual data source (network info), which would be the `WalletService`.

AdamISZ replied:
True, although it's a minor difference. I can put both added and removed as a debug msg in the `WalletService.transaction_monitor()` after `self.wallet.process_new_txs()`.

undeath replied:
Yes, I'd prefer it that way. (also: this would be the only `jlog` call in the whole wallet code)

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/248dbbad63bd27862e7b3b9354468db13d09aec3

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+        """ Encapsulates start up actions.
+        Here wallet sync.
+        """
+        Service.startService(self)
```
this would look much less confusing if it was a call to `super()`

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+        """
+        if not syncresult:
+            jlog.error("Failed to sync the bitcoin wallet. Shutting down.")
+            reactor.stop()
```
Does `reactor.stop()` immediately stop execution of the whole program? Otherwise, shouldn't there be a `return`?

AdamISZ replied:
Good catch. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+            # note that in this case, txid is ignored.
+            self.callbacks["all"].extend(callbacks)
+        elif cb_type in ["unconfirmed", "confirmed"]:
+            if txinfo not in self.callbacks[cb_type].keys():
```
`.keys()` is not needed

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+        Intended to be run as a twisted task.LoopingCall so that this
+        Service is constantly in near-realtime sync with the blockchain.
+        """
+        txlist = self.bci.list_transactions(100)
```
This looks like it's just waiting to cause trouble for someone. Only ever the 100 newest transactions are considered. bitcoind's `listtransactions` supports a `skip` parameter which could be used (which would simply be `len(self.old_txs)` if I'm not missing anything)

AdamISZ replied:
Note that we had this same logic and magic number already in the previous version (just, we were doing it per join, which is kinda yuck).

I agree that it's not a good thing, but the likelihood of it causing something to be missed is negligible (100+ transaction *updates* in parallel; something else will completely break before this), and the consequences are at worst a wallet not knowing about a tx, not a funds loss risk. My feeling is to leave it as-is for this PR, and to open an Issue to fix it to be properly generic.

undeath replied:
I'm ok with leaving an issue for that.

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+        self.synced = self.bci.wallet_synced
+        # Don't attempt updates on transactions that existed
+        # before startup
+        self.old_txs = self.bci.list_transactions(100)
```
Here `old_txs` is a list of transaction objects as returned from the rpc, but in `transaction_monitor` it is only a list of `txid`. Additionally it's not defined in `__init__`.

AdamISZ replied:
Bug! That's what caused the erroneous "new tx processing" on startup. Thanks. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/e1fbc427120e31fdc3e96fe761720abf085390f2

undeath replied:
`old_txs` is still missing from `WalletService.__init__`

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+            if x['txid'] in self.active_txids or x['txid'] not in self.old_txs:
+                new_txs.append(x)
+        # reset for next polling event:
+        self.old_txs = [x['txid'] for x in txlist]
```
For performance reasons I think this should ideally be a `set`, not a `list`, and could be updated in the loop above.

AdamISZ replied:
I've decided not to address either of these suggestions of set() to avoid further testing. I don't see any scenario currently where the performance would be noticeable. Of course I don't object to that change, though, if someone wants to PR it.

> and could be updated in the loop above.

I don't believe that's the case, since there is a comparison to the current contents of `self.old_txs` inside the loop.

undeath replied:
I don't think the comparison would cause any problems if you're adding to the set after checking it, but the change is not important anyway.

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+
+        # transactions we are actively monitoring,
+        # i.e. they are not new but we want to track:
+        self.active_txids = []
```
for performance reasons this would be better as a `set`

AdamISZ replied:
See previous

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+        for tx in new_txs:
+            txid = tx["txid"]
+            # filter on this Joinmarket wallet:
+            if "label" not in tx.keys() or \
```
`.keys()` is not needed

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5

undeath commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -0,0 +1,362 @@
+                                if f(txd, txid, confs):
+                                    self.callbacks["confirmed"][k].remove(f)
+                                    if txid in self.active_txids:
+                                        self.active_txids.remove(txid)
```
Entries from `active_txids` should also be removed if there was no corresponding callback.

AdamISZ replied:
Don't think so.

The dict `confirm_callbacks` is keyed by txinfo, either the txid expected (for the Taker, who knows it) or a tuple of outputs (for Makers, who don't , when they set up the callback), and we use active_txids as a way of keeping this transaction as one we're listening on, even after we've processed it once (at confs = 0).

If we call f(txd, txid, confs) and it returns False it means that that transaction was not applicable to that callback f, so nothing is to be done because we're still waiting for the right f to be called.

undeath replied:
I don't see any part of code where it's ensured that a callback must exist for anything added to `active_txids`. So this might lead to a memleak.

AdamISZ replied:
I see (to be clear, although I'm sure you get it: active_txids is only added to on a successful return from an unconfirm callback).

So as it stands it relies on the fact that the caller/user of WalletService adds a confirm callback matching each unconfirm callback, else the list can have dangling items where an active_txid entry is created from unconf, but no conf callback ever happens. Obviously it *is* the case, as it stands, that we always add both in JM, but they can time out and never get called. Btw a very similar thing existed in the old version of the code (there were dangling callbacks that just hung around).

Considering fixing it, the trickiness would be if you wanted to check whether any callback referenced a txid, you'd need to make sure each callback was indexed/referenced with a txid, but when they're set up, some of them aren't able to do that, but can only index by tx-out-set. They could be altered at the time of unconf callback to do that. But it is going to be a chunk of extra logic, I'm not sure it's worth doing that? (the logic of the transaction monitor is already quite nested and conditional).

To me a dangling txid in that list for every tx that doesn't get confirmed in 6 hours is basically nothing. But I agree it could be made more watertight.

undeath replied:
It's not super important. Maybe put a FIXME comment in the code somewhere and leave it alone until someone has trouble with it.

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/248dbbad63bd27862e7b3b9354468db13d09aec3

undeath commented on jmclient/test/test_coinjoin.py:
```diff
@@ -119,15 +124,17 @@ def raise_exit(i):
 
     MAKER_NUM = 3
-    wallets = make_wallets_to_list(make_wallets(
+    wss = make_wallets_to_list(make_wallets(
```
this is missing the call to the previously defined but unused `wallets_to_wallet_services`

undeath replied:
I see `make_wallets` now returns `WalletService` objects instead and even saves them in the `wallet` key. That's confusing. `wallets_to_wallet_services` can be removed?

AdamISZ replied:
Yes, exactly, sorry I just forgot to clean this up. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/e1fbc427120e31fdc3e96fe761720abf085390f2

undeath commented on scripts/sendpayment.py:
```diff
@@ -155,7 +151,18 @@ def main():
+        def wait_to_send():
+            if not ws.synced:
+                return
+            wl.stop()
```
just to make sure, does this actually work? `wl` is referenced here before being defined in the outer scope.

AdamISZ replied:
I guess this is a Python question? I've run this on regtest and it worked, yes. I guess it works because there is no context in which that code is reached without `wl` being defined?

undeath replied:
Yes, this was purely a python question. I wasn't entirely sure if would behave like that.

undeath commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -658,7 +506,7 @@ def sync_unspent(self, wallet):
         et = time.time()
         log.debug('bitcoind sync_unspent took ' + str((et - st)) + 'sec')
-        self.wallet_synced = True
+        #self.wallet_synced = True
```
is there still any good reason to keep the `wallet_synced` variable or can it be factored out completely?

AdamISZ replied:
Maybe, maybe not. For now the `sync_wallet` function in the `WalletService` still uses it to decide if it succeeded or not. As you doubtless remember that setting has a number of possible places to be set within the blockchaininterface syncing code. I do agree that it'd be nice not to have it be a separate setting at bci level.

undeath commented:
ok, I've completed my initial review and left a few comments.

I don't understand under what circumstances the `wallet` of the `WalletService` is accessed directly vs through one of `WalletService`'s proxy methods. Since it is not known to the calling code whether it is calling a pass-through method on `WalletService` or actually an intercepted method, the `wallet` should be made a protected attribute of the `WalletService` and simply all calls be routed through the `WalletService` proxy. Otherwise chaos and bugs will eventually emerge.

I'm not very confident in the `WalletService` ever actually having multi-wallet support. It'd be highly non-trivial to route every single wallet method's call to the correct wallet. But I don't think that's an important feature for joinmarket anyway.

AdamISZ commented:
Many thanks for review, it is as always extremely helpful.

> ok, I've completed my initial review and left a few comments.
> 
> I don't understand under what circumstances the `wallet` of the `WalletService` is accessed directly vs through one of `WalletService`'s proxy methods. 

It may surprise you to learn, but: neither do I! As I went through it, I think I just found in one or two places that I was uncertain whether the function should be in the WalletService interface, but I also was thinking that this should best just be tightened up and be a clean separation. I will endeavour to do that, along with the `__getattr__` trick you mention, as my next (and I guess final) step in addressing your review points.

Since it is not known to the calling code whether it is calling a pass-through method on `WalletService` or actually an intercepted method, the `wallet` should be made a protected attribute of the `WalletService` and simply all calls be routed through the `WalletService` proxy. Otherwise chaos and bugs will eventually emerge.
> 

Well, uh yes, colourfully put but I do get the point lol.

> I'm not very confident in the `WalletService` ever actually having multi-wallet support. It'd be highly non-trivial to route every single wallet method's call to the correct wallet. But I don't think that's an important feature for joinmarket anyway.

Agreed for sure. As I was trying to "sell" the encapsulation I mentioned this possibility offhand (twice I think), but I also agree it's a bad one for both the reasons of (a) being very impractical and (b) not really likely to be needed. The encapsulation *if done fully* (above paragraph) speaks for itself, at least to anyone who cares about architecture (which we must now given the size of the codebase), and more practical examples of why this PR matters at all can be found in things like my recent #403 SNICKER PR, where passive coinjoining fundamentally depends on the wallet being reactive, instead of having to be manually updated (and Qt and deposits ..). Anyway, there is nothing to discuss here, we're fully in agreement I guess.

(Oh, also, I did not address the list-> set suggestions, I guess I will do that too assuming there's no gotchas).

undeath commented:
I noticed a bunch of `dict.items()` calls, which, for performance reasons in py2, should be `iteritems(dict)` instead (from the futures lib). OTOH, isn't it time to drop py2 support?

AdamISZ commented:
> I noticed a bunch of dict.items() calls, which, for performance reasons in py2, should be iteritems(dict) instead (from the futures lib). OTOH, isn't it time to drop py2 support?

Don't think that needs to be looked at right now.

@undeath or anyone else, I have rebased locally and addressed all comments so far. Is there anything in these last commits you want to ask about? If not I will go ahead and apply the rebase with force push and move into final testing.

undeath commented:
I've reviewed all new commits and didn't spot any problems with those. Go ahead and force-push.

AdamISZ commented:
OK force push is done but as you can see I did add, after some more testing, a few more commits. The substantive part is: sweeps to external addresses were not functioning correctly (callbacks not called, depending on details), which I had not picked up in earlier tests due to running tumbler with destinations inside the same wallet.

The solution here is slightly different to the solution we had in the former version of the code: we import into the same wallet label, when we are doing a sweep. If the destination is already in the wallet, no harm done: this is a no-op import. If not though it allows listtransactions to attach it to a label we're watching for.
A `logged` variable is also added to avoid repeated re-logging of the same utxos. It's a bit ugly but only in a trivial way, and the alternatives were too invasive or complicated.

A similar detail was that Taker scripts cannot start up properly without the wallet synced in certain cases (e.g. sweeps), so simply forced the sync of the wallet to happen upfront always.

One last commit I mustn't forget is to remove the imports of 'collect_addresses_gap' which crept in here, since it was explicitly removed in #402 because of a horrible Core bug affecting slow hard drives (see earlier comments about how we'll have to address this later, but it should be easier with this Service arch).

Just going to do that now with one commit and then start more live testing, others please do the same if you have a chance. Will squash but only just before merge.

AdamISZ commented:
https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/d875fe78f59efbd7803d6434b17b5d0a4b39d39e is a small but important change - I'd erroneously assumed it was always fine to import external addresses to our wallet label, but it isn't - it breaks fast sync, which assumes that any used addresses in the label belong to this wallet, and since we switched to fast sync as the default, that's obviously wrong, hence these few lines added make it so that non-wallet imports go to `joinmarket-notify` (as they did before).

The other changes are minor and self-explanatory. I will continue doing testing (have done a bit of mainnet testing already, everything is working well so far apart from those points.

kristapsk commented:
Found an issue with Qt. When you open a wallet, correct balances are not shown initially, it's 0.00000000 everywhere. Then after few seconds, it updates to correct ones. That can be confusing, at least we should have empty strings not zeroes or some spinning wheel or something like that, if we can't show correct values immediately (haven't looked at the code much yet).

AdamISZ commented:
> Found an issue with Qt. When you open a wallet, correct balances are not shown initially, it's 0.00000000 everywhere. Then after few seconds, it updates to correct ones. That can be confusing, at least we should have empty strings not zeroes or some spinning wheel or something like that, if we can't show correct values immediately (haven't looked at the code much yet).

Yes, thanks for prompting me on this, it's one of I think 4 minor issues with how the Qt interface looks:

1. balance showing 0 before sync completes
2. item selection (highlight) can be lost after a few seconds
3. very slight slow responsiveness sometimes
4. Information text at the bottom (quite useful) is getting overwritten with "Wallet synced" every few seconds.

Clearly, even though it only lasts at most a few seconds, 1, which you highlighted, is not acceptable because of what it implies to the user, so I have to fix that before merging really. The other three are fairly minor. They need to be addressed, but not necessarily right now.
I hope people will notice the considerable improvements: expansion states are remembered in the wallet tab so it's far easier to navigate, and the wallet updates itself automatically, no un-obvious manual syncing needed, including for deposits from outside as well as coinjoins.

AdamISZ commented:
Pushed fix for above, the other 3 points may still be minor irritants, depending. Tests OK here.

AlexCato commented:
Might have found a problem:
When a coinjoin happens, the offers for the involved makers are created new. This PR seems to also consider 0-conf UTXOs to be available for joins.

Example:
Mixdepth0=1 btc currently, mixdepth1=2btc. If a join with 0.5 btc happens (using mixdepth0 to fill), it immediately offers a new max of 2.5 BTC (since those 0.5 moved from mix0 to mix1) when the transaction is seen on the network, but BEFORE confirmation of it.

AdamISZ commented:
@AlexCato just to ping you that I answered in some detail on IRC, I am trying out one approach based on that, but not sure, further input based on that would be appreciated. I could paste it here if it helps.

AdamISZ commented:
I mean we could but it just felt like taking up too much space.

chris-belcher commented:
The space used is really worth it for the readability and maintainability. The code would become more self-documenting and easier to understand for a maintainer who could be working years from now.

AlexCato commented:
Second @chris-belcher here, I also prefer self-explanatory variable names and dont care if the code is a bit longer.

I've seen your explanation on IRC Adam, here it is for others to see:

> waxwing_] so to explain the situation, it's like this:
> waxwing_] in the current master (and earlier), we remove utxos when the tx is seen on the network (so unconf), but don't add the new ones. when confirmation happens, we do wallet.sync_unspent.
> waxwing_] which gives rise to a temporary period where we're offering less than we "have", but we don't advertise unconfirmed coins for join. however, there is `listunspentargs` which means that when we select for a join, we can select unconfirmed if that's set to [0]
> waxwing_] no hang on i think that last statement is wrong, let me check.
> waxwing_] yeah it's wrong because of the manual changes to our wallet's utxos; we first manually remove the spent ones at broadcast time, so obviously not offering new unconf utxos then, and then we do sync_unspent *only* when the confirmation has happened (sync_unspent uses listunspent_args), which means whatever the setting is doesn't matter; those coins are now confirmed.
> waxwing_] so in the new version, we basically process every transaction notification and add and remove utxos when they're seen, and that's happening under the hood; and callbacks happen if a callback is attached to a particular transaction.
> waxwing_] reference: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/files#diff-963259b630dee1562e2bcaa5de98e960R211
> waxwing_] so what to do, hmm, well perhaps simplest is just have a flag to process_new_tx and when we reach that line ^ in the unconf case, we just pass the flag and it only removes, but does not add.
> waxwing_] however i thought about that and it does have a bit of a drawback; it means coins don't get seen when they happen on the network.
> waxwing_] at the moment, if i deposit into my JMQt wallet i see the coins within a couple of seconds (i mean, with #359 ofc)
> waxwing_] that was kind of the main point (and it'll help with other features, like snicker, or basically anything that wants to use the wallet without manually interacting with it, so to speak)
> waxwing_] i think the only full answer is to attach confirmation metadata to the utxos.

One note from my side here:
This needs to work differently for wallet-tools (for viewing) and makers. I remember a few years back, only confirmed coins were shown everywhere. People had heart attacks when they first tried out joinmarket, when they sent coins to their wallet which didnt show up. So desired behavior is: wallet-tools also show unconfirmed TX while makers/takers only use confirmed ones for any joins (but remove unconfirmed, but already-seen-on-network-TX from being used for other joins until timeout).

AdamISZ commented:
> Second @chris-belcher here, I also prefer self-explanatory variable names and dont care if the code is a bit longer.

OK.

> This needs to work differently for wallet-tools (for viewing) and makers

Yes, I hadn't forgotten that point. I have a commit now that adds confirmation info to the utxos in the wallet, and have set it so the defaults for get_utxos_by_mixdepth and get_balance_by_mixdepth are to not change behaviour (show all) but only have makers and takers change their behaviour for joins (both in selecting for txs and for making offers maker side). Direct send I set as allowing unconfs.
Note that wallet display (in wallet-tool or Qt which both use wallet_display) behaviour can still be changed by resetting `listunspent_args` because that still controls what comes out of `sync_unspent`.

kristapsk commented:
Have done some testing on testnet last few days (no coinjoins, just simple wallet opening, sending, receiving), haven't found any new issues.

Will apply this to mainnet yg and test more when the 0conf cj issue mentioned above is addressed.

AdamISZ commented:
@kristapsk thanks for that. And of course thanks @AlexCato @chris-belcher @undeath for reviewing and testing this, it turns out to be a bit more involved than I expected, not surprising in retrospect due to reworking the inner mechanics of the interaction between the blockchain and the application.

AdamISZ commented:
So https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/c5d8d2614a50ceadda1fe9ab8de3b700ea96b3fb turned out to be a BIG update/change, as a result of AlexCato's observation on selection of unconfirmed coins. In order to make all the parts fit together nicely, I ended up needing to do all this:

* The underlying utxo set in the wallet (`UTXOManager._utxo`) has entries which now include a "height" field. This is set to pseudo-infinity while unconfirmed and set to the actual block height in which it is included, once confirmed. The reason for using block height rather than confirmations is, of course, that if the latter, it would need to be constantly updated with each passing block. This is fairly simple to do though with simple rpc calls.

Having done this we now can update the methods `get_utxos_by_mixdepth`, `get_balance_by_mixdepth` and `select_utxos` to filter on a specific number of confirmations. It passes through two layers - calling applications (like `jmclient.Taker`), which know nothing about block heights, can ask the `WalletService` object they own for utxos with at least X confirmations, and then the WalletService object wraps the corresponding function in the Wallet object, doing the translation between confirmations and absolute block height, passing that block height to the internal function which queries the `UTXOManager` object owned by the Wallet. The WalletService can do this because it updates the latest block height within its heartbeat main loop. (This can lead to being out of sync on how many confs by a few seconds, but this should not be a problem at all).

Once that's all set up, we can then easily make these observations/stipulations:

* Direct send allows spending any coins (of course this can and should be made optional for the user in future, but for now unconf spending is the default here)
* Taker sweep spends all the coins, confirmed or not. It would be unintuitive to behave otherwise.
* Taker normal one-shot coinjoin filters on 1 confirmation. Feel free to disagree with me on that (see below 'antisocial' comment).
* For tumbler serial transacting, it's not an issue since we only start the next tx after confirm + wait anyway
* For commitment finding (PoDLE), it's not relevant as we explicitly filter the confirmations after
* Maker offer calculation (`get_balance_by_mixdepth` in yieldgenerator) will require 1 conf (this is the direct addressing of @AlexCato 's earlier observation).
* Maker in normal coinjoin running will only select utxos with at least 1 conf (this is the kind of "don't be antisocial" principle which we have tended to operate on, but have not been too explicit or consistent about .. and of course it must be consistent with previous bullet point).
* Wallet display shows all coins confirmed or not. (Future update to show how many confs on command line and Qt should now be simpler, in particular in the latter case it will update dynamically without further code, once it's displayed).

Further to this, I've now removed all the wallet syncing logic from `blockchaininterface.py`; most egregiously, as mentioned by @undeath (and I've winced about it several times in the past), having the wallet synced state stored in the BlockchainInterface object was a mess, as was the parsing of fast/non fast. So a lot of redundant code was further removed from blockchaininterface.py, which is now only a fairly simple shell around querying state such as getting transaction, utxo information, getting blocks etc. However what was moved was largely kept intact: the `sync_unspent`, `sync_address`, `get_address_usages` functions were copied wholesale into the `WalletService` with only syntax changes rather than semantic ones. Now the wallet `synced` state is stored there only.

The remainder of the changes are to tests. Perhaps surprisingly the `test_blockchaininterface.py` needed few changes, as it was mainly testing syncing modes, which don't change from the top level, with this commit.

AdamISZ commented:
After https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/b186c4c3042900f1440c8286c9564c64db7e315d this is ready for mainnet testing again.

kristapsk commented:
Tried to apply this to my mainnet yg, had issues that even wallet-tool thinks I have zero balance in wallet. Also, yg outputs "You do not have the minimum required amount of coins to be a maker" error and "Failed to create offers, giving up." message (obviously), followed by an unhandled exception. Will try to look into what's happening there in coming days.

AlexCato commented:
After running the new commits on a maker for a few days, I can confirm that it works perfectly fine (seen a few joins) and that this is fixed:

> Might have found a problem:
> When a coinjoin happens, the offers for the involved makers are created new. This PR seems to also consider 0-conf UTXOs to be available for joins.

chris-belcher commented:
I tried to recover a wallet only from the seed phrase (so brand new wallet.dat, import every address and rescan). The importing of addresses resulted in this crash.

```
2019-10-22 00:16:47,997 [DEBUG]  rpc: listaddressgroupings []
2019-10-22 00:16:48,023 [INFO]  Detected new wallet, performing initial import
2019-10-22 00:16:48,024 [DEBUG]  requesting detailed wallet history
2019-10-22 00:16:54,662 [DEBUG]  rpc: getaddressesbyaccount ['joinmarket-wallet-745fa0']
2019-10-22 00:16:54,747 [DEBUG]  rpc: listlabels []
Traceback (most recent call last):
  File "wallet-tool.py", line 13, in <module>
    jmprint(wallet_tool_main("wallets"), "success")
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/wallet_utils.py", line 1221, in wallet_tool_main
    if wallet_service.sync_wallet(fast = not options.recoversync):
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/wallet_service.py", line 328, in sync_wallet
    self.sync_wallet_fast()
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/wallet_service.py", line 352, in sync_wallet_fast
    self.get_address_usages()
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/wallet_service.py", line 380, in get_address_usages
    self.sync_addresses()
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/wallet_service.py", line 458, in sync_addresses
    wallet_name)
  File "/home/username/coding/joinmarket/joinmarket-clientserver-2/jmclient/jmclient/blockchaininterface.py", line 235, in add_watchonly_addresses
    if self.restart_cb:
AttributeError: 'BitcoinCoreInterface' object has no attribute 'restart_cb'
```

AdamISZ commented:
@chris-belcher @AlexCato thanks. @kristapsk - It's been a while, have you checked again and specifically were you running dad8bcc ?

chris-belcher commented:
Yes I was running that commit.

AdamISZ commented:
@chris-belcher I think you misread :) I wasn't asking you that, only @kristapsk specifically - and that, because what he described on IRC at the time led to me to suspect that a code divergence may have caused the crash he experienced. For your case, I don't find it surprising that there may be such an error considering I haven't re-tested manual recovery since the latest changes, so I'll just repro and fix it today, most likely.

AdamISZ commented:
So https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/3a60c6759a1c4db3c29f711517d9516a2bfc8758 should fix that.

It came from me moving code around while refactoring. Now it behaves as before: if there is no restart callback specified, it defaults to command line printout. The bug wasn't specific to recovery (as is I guess obvious).

A reminder that `--recoversync` should be used when recovering on a new instance (as above, way above, in the scrollback, we now use fast as default and what was default, is called 'recoversync').

chris-belcher commented:
With that new commit it correctly recovered a mainnet wallet from seed phrase.

kristapsk commented:
@AdamISZ , no, was busy with other stuff and now I'm travelling this week, so cannot promise anything in coming days.

AdamISZ commented:
@kristapsk understood; thanks for letting me know.

AdamISZ commented:
@kristapsk fwiw tested `summary` and seems fine. And I have one more comment after our long discussion yesterday: on the one hand you said that `showutxos` returned `{}`, but a little further on you said, that you saw one address with coins in mixdepth 4. That's particularly weird, surely something conflicts there. I still don't have good ideas, maybe try to recover the wallet with a new rpc_wallet_file and see if that fixes it. So far I haven't found a reproduction of your problem (or a way to do it), including @AlexCato saying he is able to sync large/old wallets.
